src/modules/persistent-store-base.ts

```diff
@@ -0,0 +1,76 @@
+import { Store } from 'store-unit';
+
+type Options<S> = {
+  retrieve: (key: string) => Promise<S | undefined>;
+  save: (key: string, value: S) => Promise<void>;
+};
+
+export class PersistentStoreBase<T> extends Store<T> {
+  protected key: string;
+  protected isReady: boolean;
+  private readyPromise: Promise<void>;
+
+  public options: Options<T>;
+  public defaultOptions: Options<T> = {
+    retrieve: () => {
+      throw new Error('PersistentStore:retrieve Not Implemented');
+    },
+    save: () => {
+      throw new Error('PersistentStore:save Not Implemented');
+    },
+  };
+
+  constructor(initialState: T, key: string, options: Partial<Options<T>> = {}) {
+    super(initialState);
+    this.key = key;
+    this.options = { ...this.defaultOptions, ...options };
+    this.isReady = false;
+    this.readyPromise = this.restore();
+    this.on('change', (state) => {
+      // only write to disk after restore so that we're not making
+      // a redundant write on each initialization
+      if (this.isReady) {
+        this.options.save(this.key, state);
+      }
+    });
+  }
+
+  getState() {
+    if (!this.isReady) {
+      throw new Error('Do not access getState() before checking ready()');
+    }
+    return super.getState();
+  }
+
+  setState(...args: Parameters<Store<T>['setState']>) {
```

Comment:
Should we specify setState as async here?

Reply:
It wouldn't change the type though

Reply:
Yes! But will make it a bit more obvious :)